src/modules/audio/spatializers/oculus_spatializer.c

```diff
@@ -0,0 +1,222 @@
+  bool occupied; // If true either source->playing or Oculus Audio is doing an echo tailoff
+} SourceRecord;
+
+struct {
+  uint32_t sampleRate;
+
+  ovrAudioContext context;
+  SourceRecord *sources;
+  int sourceMax;   // Maximum allowed simultaneous sources
+
+  int sourceCount; // Number of active sources seen this playback
+  int occupiedCount; // Number of sources+tailoffs seen this playback (ie strictly gte sourceCount)
+  bool midPlayback; // An onPlayback callback is in progress
+
+  bool poseUpdated; // setListenerPose has been called since the last playback
+  ovrPoseStatef pose;
+  ma_mutex poseLock; // Using ma_mutex in case holding a lovr lock inside a ma lock is weird
+  bool poseLockInited;
+} state;
+
+static bool oculus_spatializer_init(SpatializerConfigIn configIn, SpatializerConfigOut *configOut)
+{
+  // Initialize own state
+  state.sampleRate = configIn.sampleRate;
+  configOut->needFixedBuffer = true;
+  state.sourceMax = configIn.maxSourcesHint;
+  state.sources = calloc(state.sourceMax, sizeof(SourceRecord));
+
+  if (!state.poseLockInited) {
+    int mutexStatus = ma_mutex_init(&state.poseLock);
+    lovrAssert(mutexStatus == MA_SUCCESS, "Failed to create audio mutex");
+    state.poseLockInited = true;
+  }
+
+  // Initialize Oculus
+  ovrAudioContextConfiguration config = {0};
+
+  config.acc_Size = sizeof( config );
+  config.acc_MaxNumSources = state.sourceMax;
+  config.acc_SampleRate = state.sampleRate;
+  config.acc_BufferLength = configIn.fixedBuffer; // Stereo
+
+  if ( ovrAudio_CreateContext( &state.context, &config ) != ovrSuccess )
+  {
+    return false;
+  }
+
+  return true;
+}
+static void oculus_spatializer_destroy(void)
+{
+  free(state.sources);
+  state.sources = NULL;
+}
+static uint32_t oculus_spatializer_source_apply(Source* source, const float* input, float* output, uint32_t framesIn, uint32_t framesOut) {
+  if (!state.midPlayback) { // Run this code only on the first Source of a playback
+    state.midPlayback = true;
+
+    for(int idx = 0; idx < state.sourceMax; idx++) { // Clear presence tracking and get starting positions
+      SourceRecord *record = &state.sources[idx];
+      record->usedSourceThisPlayback = false;
+      if (record->source)
+        state.sourceCount++;
+      if (record->occupied)
+        state.occupiedCount++;
+    }
+
+    if (state.poseUpdated) {
+      { // Tell Oculus Audio where the headset is
+        ovrPoseStatef pose;
+
+        ma_mutex_lock(&state.poseLock); // Do nothing inside lock but make a copy of the pose
+        memcpy(&pose, &state.pose, sizeof(pose));
+        state.poseUpdated = false;
+        ma_mutex_unlock(&state.poseLock);
+
+        ovrAudio_SetListenerPoseStatef(state.context, &pose); // Upload pose
+      }
+      state.poseUpdated = false;
+    }
+  }
+
+  intptr_t *spatializerMemo = lovrSourceGetSpatializerMemoField(source);
+

```

Comment:
I apologize for the complexity of this next section-- it's just plain doing something complicated. It's actually simpler than the first two versions…!
Things to note: Aside from the tail-killing rule it doesn't try to be "fair", if you play too many sounds at once it mutes the sources that haven't nabbed record fields yet, then they will suddenly (and unfairly) unmute as other sources stop.
Note this is all complicated and therefore a likely site for bugs, but I haven't stress-tested it yet.